cdap-explore/src/main/java/co/cask/cdap/explore/service/hive/BaseHiveExploreService.java

```diff
@@ -1135,8 +1174,47 @@ private String getHiveDatabase(@Nullable String namespace) {
     if (namespace == null) {
       return null;
     }
-    String tablePrefix = cConf.get(Constants.Dataset.TABLE_PREFIX);
-    return namespace.equals(Id.Namespace.DEFAULT.getId()) ? namespace : String.format("%s_%s", tablePrefix, namespace);
+    try {
+      if (namespace.equals(Id.Namespace.DEFAULT.getId())) {
+        return namespace;
+      } else {
+        String hiveDatabase = getCustomHiveDatabaseIfExists(namespace);
+        if (!Strings.isNullOrEmpty(hiveDatabase)) {
+          return hiveDatabase;
+        } else {
+          String tablePrefix = cConf.get(Constants.Dataset.TABLE_PREFIX);
+          return String.format("%s_%s", tablePrefix, namespace);
+        }
+      }
+    } catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Return the custom database for a given namespace if one exists else return null
+   *
+   * @param namespace the namespace for which we need to look up custom mapping
+   * @return the custom hive database name if one exists else null
+   * @throws ExploreException if failed to get NamespaceMeta for the given namespace
+   */
+  @Nullable
+  private String getCustomHiveDatabaseIfExists(@Nullable String namespace) throws ExploreException {
```

Comment:
actually, this method is a little strange. the logic in this method depends on a parameter being null. Also, the logic in the methods calling this depends on the return value from this being null. Too many nulls :) 

If this method is called supposed to ``getCustomHiveDatabase``, then we should perhaps only be called when there is custom mapping for hive.

WIthout that, you have the same check on line 1212 in this method, and line 798 in the caller.